### PR TITLE
Improved memory usage when generating large MARC files.

### PR DESCRIPTION
**Pull request by aslagle**
This branch uses multipart S3 uploads and batched queries to reduce the memory usage when creating MARC files for large collections.

I chose the batch sizes by testing with a large local collection, and the maximum memory use I got from this was around 3GB, which is still a lot but it was well over 10GB before. @jonathangreen does that seem good enough?

### CODE REVIEW

File: marc.py [hunk 1]
```diff
@@ -592,52 +592,91 @@ def create_record(cls, work, annotator, force_create=False, integration=None):
...
+        offset = 0
+
+        url = mirror.marc_file_url(self.library, lane, end_time, start_time)
+        representation, ignore = get_one_or_create(
+            self._db, Representation, url=url, media_type=Representation.MARC_MEDIA_TYPE)
+
+        with mirror.multipart_upload(representation, url) as upload:
+            output = StringIO()
+            current_count = 0
+            while offset < total:
```
Comment (leonardr) [inline marc.py]: I've spent some time mentally fiddling around with the idea of turning this into a generator and passing the generator into `mirror`. But what I'm actually thinking is that I'd rather see the code inside this loop in a helper method that can be tested separately.

Your existing test coverage is fine, so don't worry aobut this.

File: marc.py [hunk 2]
```diff
@@ -592,52 +592,91 @@ def create_record(cls, work, annotator, force_create=False, integration=None):
...
+
+        # We mirror the content, if it's not empty. If it's empty, we create a CachedMARCFile
+        # and Representation, but don't actually mirror it.
+        if not mirror:
+            storage_protocol = self.integration.setting(self.STORAGE_PROTOCOL).value
+            if storage_protocol == ExternalIntegration.S3:
+                mirror_integration = ExternalIntegration.lookup(
+                    self._db, storage_protocol, ExternalIntegration.STORAGE_GOAL)
+                if mirror_integration:
+                    mirror = mirror or S3Uploader(mirror_integration)
```
Comment (leonardr) [inline marc.py]: This seems very similar to `MirrorUploader.sitewide`. Can you use that instead?

Comment (aslagle) [reply marc.py]: I think we talked about this in my first MARC PR too. I had originally thought of having an option to store MARC files on disk so it wasn't necessary to set up S3, but I'm not sure if anyone will need that. I didn't like the assumption that there would only ever be one storage option configured though.

Comment (leonardr) [reply marc.py]: When we support more than one storage option, this code will probably need to be changed anyway. This code implies that there can be multiple storage options, but only one per protocol.

Not a big deal -- we can revisit this when we add other storage options. My only concern is that when that happens I will probably go through the code looking for references to sitewide() and I might miss this one.

**Comment by jonathangreen** [comment]
@aslagle 3gb seems doable to me, much better than it was before. 👍🏻

**Pull request merged**